server/src/main/java/com/google/fhir/gateway/BundleProcessor.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright 2021-2023 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.fhir.gateway;
+
+import ca.uhn.fhir.context.FhirContext;
+import ca.uhn.fhir.parser.IParser;
+import ca.uhn.fhir.rest.server.exceptions.InvalidRequestException;
+import com.google.fhir.gateway.interfaces.RequestDetailsReader;
+import java.nio.charset.Charset;
+import java.nio.charset.StandardCharsets;
+import java.util.Map;
+import java.util.function.Consumer;
+import java.util.function.Supplier;
+import org.hl7.fhir.instance.model.api.IBaseResource;
+import org.hl7.fhir.r4.model.Bundle;
+import org.hl7.fhir.r4.model.Bundle.BundleEntryComponent;
+import org.hl7.fhir.r4.model.Bundle.BundleType;
+import org.hl7.fhir.r4.model.Bundle.HTTPVerb;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class BundleProcessor {
+  private static final Logger logger = LoggerFactory.getLogger(BundleProcessor.class);
+
+  private final FhirContext fhirContext;
+
+  public BundleProcessor(FhirContext fhirContext) {
+    this.fhirContext = fhirContext;
+  }
+
+  public <R> R processBundleFromRequest(
+      RequestDetailsReader request,
+      Map<HTTPVerb, Consumer<BundleEntryComponent>> entryComponentProcessors,
+      Supplier<R> postEntriesProcessor) {
```

Comment:
It took me a while to understand the pattern you have designed here. While I think this is interesting, I feel it is too complicated. A couple of suggestions:

1) Let's first define the role of `BundleProcessor`. To me, it should be just for parsing a Bundle request and call a given worker for each entry. In future, we may need to be able to chain workers but for now the need is not there.

2) For that role, the interface of this method can simply be:
```
public void processBundleFromRequest(
    RequestDetailsReader request,
    Consumer<BundleEntryComponent> entryProcessors)
```
So in other words, leave the mapping from `HTTPVerb` to methods to the supplied `entryProcessor`. All the access related logic/state should be handled by the `entryProcessor`.

3) What is the benefit of `postEntriesProcessor` here? Is it just for convenience of avoiding one call in the user code? I feel the complexity it is adding (especially around the generic `R`) does not worth the benefit. So my suggestion is to leave that out completely. So for example, in the case of `PatientAccessChecker` we simply call `areAllBundleOperationsAllowed` after `processBundleFromRequest` is done. We can do a similar pattern with `PatientFinderImp`.

4) I think it is a useful functionality to stop processing early. For example in an access-checker we may want to stop entry processing as soon as we find the first violation. So maybe instead of `Consumer<BundleEntryComponent> entryProcessors` define a custom interface that returns true/false to decide whether to continue or not. On that note, is there a reason you don't do that in `PatientAccessChecker` (I have asked above too)?

Reply:
I am having a hard time understanding the pattern. 
I am not able to see a lot of value in this new abstraction that is added and feel this is very complicated. 

Also the intent was to simplify code duplication in bundle processing. I am not able to see that happening. Please correct me if this is not the case.

Reply:
Yes, I agree the new implementation is still too complicated; in fact, I was not expecting this much change for doing this refactoring. Let's step back and see why we are doing this; there are two main reasons:
1) performance
2) code duplication in Bundle processing

We should definitely fix (1), i.e., we should not parse the payload twice to create the same Bundle. To fix this (and only this) we can simply have a helper method that creates a `Bundle` from the request payload. Let's first do this. So `PatientFinder.findPatientsInBundle` will receive a `Bundle` instead of a `RequestDetailsReader` (and conversion of the payload to Bundle happens outside that method). The same `Bundle` is also used in `PatientAccessChecker` where we want to check the scopes.

Does this sound good to everyone?

If yes, let's start with this only, which mostly addresses (1). We can do (2) in a separate PR if we still see the code duplication is significant. But I expect the first PR for addressing (1) should be much simpler than the current version.

Reply:
Sure, I agree. We can just focus on the first aspect as @bashir2  mentioned.
Since, there are already a lot of changes that have been made to this PR, for the suggested changes I believe it will be easier to start from the `main` branch. Hence will close this PR